src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt

```diff
@@ -0,0 +1,57 @@
+package uk.gov.justice.digital.hmpps.incentivesapi.jpa.repository
+
+import kotlinx.coroutines.flow.Flow
+import org.springframework.data.r2dbc.repository.Query
+import org.springframework.data.repository.kotlin.CoroutineCrudRepository
+import org.springframework.stereotype.Repository
+import uk.gov.justice.digital.hmpps.incentivesapi.jpa.Kpi
+import java.time.LocalDate
+
+data class ReviewsConductedPrisonersReviewed(
+  val reviewsConducted: Int,
+  val prisonersReviewed: Int,
+)
+
+data class PrisonerNumberOverdue(
+  val prisonerNumber: String,
+)
+
+@Repository
+interface KpiRepository : CoroutineCrudRepository<Kpi, LocalDate> {
+  /**
+   * Returns the number of reviews conducted and prisoners reviewed in the previous month
+   */
+  @Query(
+    //language=postgresql
+    """
+    SELECT count(*) AS reviews_conducted, count(DISTINCT prisoner_number) AS prisoners_reviewed
+    FROM prisoner_iep_level
+    -- review time between 1st day of previous month and 1st day of current month (relative to `day` argument)
+    -- e.g. day = 2023-10-19, review_time >= 2023-09-01 AND review_time < 2023-10-01
+    WHERE review_time >= (DATE_TRUNC('month', :day)-'1 month'::INTERVAL)::DATE AND
+          review_time < DATE_TRUNC('month', :day)::DATE AND
+          review_type IN ('REVIEW', 'MIGRATED')
+    ;
+    """,
+  )
+  suspend fun getNumberOfReviewsConductedAndPrisonersReviewed(day: LocalDate): ReviewsConductedPrisonersReviewed
+
+  /**
+   * Returns the prisoner numbers overdue a review
+   *
+   * NOTE: These people may no longer be in prison. To get the number of overdue *prisoners* filter out people no longer
+   * in prison.
+   */
+  @Query(
+    //language=postgresql
+    """
+    -- this may include people no longer in prison
+    SELECT DISTINCT ON (prisoner_number) booking_id, prisoner_number
+    FROM prisoner_iep_level
+    JOIN next_review_date USING (booking_id)
+    WHERE next_review_date < CURRENT_DATE
+    ;
```

Comment:
with `SELECT DISTINCT ON (…)` you need to have an `ORDER BY` clause so as to pick the latest review rather than _some_ review:

`ORDER BY prisoner_number, review_time DESC NULLS LAST`

Reply:
Thanks for the explanation earlier Igor. I'm still not 100% sure the ORDER by would make a difference in our case (given we're only interested in the number of overdue prisoners) but I've added just to be on the safe side.

Also good to have a well-defined order given how `DISTINCT ON` work (e.g. it is a glorified group by).

I've also added `review_time` to the `SELECT`, even tho we don't use that data. But the order of the rows returned seem to change so I've deemed to be saver to add that too.

If the query still doesn't look right let me know: https://github.com/ministryofjustice/hmpps-incentives-api/pull/503/commits/aae18d5caca95a4dbdd3a3ef57e3c17e31945181

Reply:
True. Although the little class may be more open for extension in the future. But I'll probably change.

Reply:
honestly, i'm happy with a named object!

Reply:
Hopefully this commit message make some sense 😅: [: Fix query to get overdue prisoners](https://github.com/ministryofjustice/hmpps-incentives-api/pull/503/commits/4896983ef9babbab0ecf9db0e2e493e5820bebef)

I tweaked the test so that the data had that edge case where a single prisoner number had multiple bookings, some with overdue review but some not overdue.

Reply:
I've also simplified the return type of that repository method: 7ab3544fe5f049145caf12158079a9e42d0b5835